File: py/Languages/Python/PythonLanguage.py

from PySide6.QtGui import QSyntaxHighlighter, QTextCharFormat, QFont
from PySide6.QtCore import QRegularExpression, Qt

from ..Language import Language, ClassDef, MethodDef, FunctionDef

from ..ASTPatterns import OptionalNode, NodeSequence, RepeatingNode
from ..ASTPatterns import NodeBranch, LateDefinedASTPattern

from ..Tokens import Token, KeywordsTokenMatcher, RegexMatcher
from ..Tokens import DirectTokenMatcher, LiteralTokenMatcher, TokenNodePattern

from ..SemanticASTNodes import CodeBlock, ImportNode, AsImportNode

from collections import OrderedDict

class PythonLanguage(Language):
    
    def name(self):
        return "Python"
    
    def tokenMatchers(self): # list<TokenMatcher>
        return [
            KeywordsTokenMatcher([
                "from", "import", "assert",
                "def", "pass", "class", "lambda",
                "if", "elif", "else", "not",
                "return", "break", "continue", 
                "raise", "except", "try", "finally",
                "while", "for",
                "True", "False", "None",
                "and", "or", "xor", "not",
                "del", "as", "in", "breakpoint"
            ], self),
            #RegexMatcher(r'\n+', "T_WHITESPACE"),
            RegexMatcher(r'\n( +)?', "T_INDENTATION"),
            #RegexMatcher(r'\s+', "T_WHITESPACE"),
            RegexMatcher(r'\s+', "T_WHITESPACE"),
            DirectTokenMatcher("@", "T_DECORATOR"),
            RegexMatcher(r'[a-zA-Z_][a-zA-Z0-9_]+', "T_SYMBOL"),
            RegexMatcher(r'[0-9_]+(\.[0-9]*)?', "T_NUMBER"),
            RegexMatcher(r'\#([^\n]*)', "T_COMMENT"),
            DirectTokenMatcher([
                ".", ",", "(", ")", "[", "]", "{", "}", ":", "="
            ], "T_SPECIAL_CHAR"),
            DirectTokenMatcher(
                ["+", "-", "*", "/"], 
                "T_OPERATOR"
            ),
            LiteralTokenMatcher("'", "T_LITERAL"),
            LiteralTokenMatcher('"', "T_LITERAL")
        ]

    def isNodeRelevantForGrammar(self, node): # boolean
        if isinstance(node, Token):
            if node.tokenName in ["T_WHITESPACE", "T_COMMENT", "T_INDENTATION"]:
                return False
        return True

    def grammar(self):
        
        #return [NodeSequence("test", [
        #    TokenNodePattern("T_FROM"),
        #    TokenNodePattern("T_SYMBOL")
        #])]
        
        # return [RepeatingNode("test", TokenNodePattern("T_INDENTATION"), False)]
        
        expression = LateDefinedASTPattern()

        identifier = NodeSequence("identifier", [
            TokenNodePattern("T_SYMBOL"),
            RepeatingNode("identifier-path", NodeSequence("identifier-element", [
                TokenNodePattern("."),
                TokenNodePattern("T_SYMBOL")
            ]), True)
        ])
        
        importFrom = NodeSequence("import-from", [
            TokenNodePattern("T_FROM"),
            RepeatingNode("import-from-parent", TokenNodePattern("."), True),
            identifier
        ])

        importStatement = NodeSequence("import", [
            OptionalNode(importFrom),
            TokenNodePattern("T_IMPORT"),
            identifier
        ])
        
        tuple = NodeSequence("tuple", [
            TokenNodePattern("("),
            RepeatingNode("tuple-content", NodeSequence("tuple-element", [
                expression,
                OptionalNode(TokenNodePattern(","))
            ]), True),
            TokenNodePattern(")")
        ])
        
        call = NodeSequence("call", [
            identifier,
            tuple
        ])
        
        raiseDef = NodeSequence("raise", [
            TokenNodePattern("T_RAISE"),
            identifier,
            tuple
        ])
        
        decorator = NodeSequence("decorator", [
            TokenNodePattern("@"),
            identifier,
            OptionalNode(tuple)
        ])
        
        operation = NodeSequence("operation", [
            expression,
            TokenNodePattern("T_OPERATOR"),
            expression
        ])
        
        assignment = NodeSequence("assignment", [
            expression,
            TokenNodePattern("="),
            expression
        ])
        
        classDefinition = NodeSequence("class", [
            TokenNodePattern("T_CLASS"),
            identifier,
            tuple,
            TokenNodePattern(":"),
        ])
        
        functionDefinition = NodeSequence("function", [
            TokenNodePattern("T_DEF"),
            identifier,
            tuple,
            TokenNodePattern(":"),
        ])
        
        expression.definePattern(NodeBranch("expression", [
            identifier,
            call,
            TokenNodePattern("T_LITERAL"),
            TokenNodePattern("T_NUMBER"),
            # operation # Infinite loop? expr => operation => expression => ... HOW?!
        ]))
        
        return [
            importStatement,
            importFrom,
            identifier,
            tuple,
            classDefinition,
            functionDefinition,
            raiseDef,
            call,
            decorator,
            operation,
            assignment,
            expression
        ]

    def groupStatementsIntoBlocks(self, nodes): # list<ASTNode>
        level = 0
        currentIndentation = 0
        stack = [CodeBlock(self, 1, 1, 0)]
        depthStack = [0]
        for node in nodes:
            if isinstance(node, Token) and node.tokenName == "T_INDENTATION":
                currentIndentation = node.code.count(" ")
                if currentIndentation > level:
                    level = currentIndentation
                    stack.append(CodeBlock(self, node.row, node.col, node.offset))
                    depthStack.append(level)
                elif currentIndentation < level:
                    while currentIndentation < level:
                        endedBlock = stack.pop()
                        depthStack.pop()
                        level = depthStack[-1]
                        stack[-1].addStatement(endedBlock)
            stack[-1].addStatement(node)
        while len(stack) > 1:
            block = stack.pop()
            stack[-1].addStatement(block)
        return stack[0].children

    def formatForNode(self, node):
        if type(node) == Token: 
            format = QTextCharFormat()

            if node.tokenName in ["T_CLASS", "T_DEF"]:
                format.setFontWeight(QFont.Bold)
                format.setForeground(Qt.darkRed)

            if node.tokenName in [
                "T_IF", "T_ELSE", "T_ELIF", 
                "T_PASS", "T_RAISE", "T_RETURN", 
                "T_BREAK", "T_CONTINUE",
                "T_ASSERT",
                "T_WHILE", "T_FOR",
                "T_TRY", "T_CATCH", "T_EXCEPT",
                "T_AND", "T_OR", "T_XOR", "T_NOT"
            ]:
                format.setFontWeight(QFont.Bold)
                format.setForeground(Qt.darkMagenta)

            if node.tokenName in ["T_NONE", "T_TRUE", "T_FALSE"]:
                format.setForeground(Qt.magenta)

            if node.tokenName in ["T_FROM", "T_IMPORT"]:
                format.setForeground(Qt.darkYellow)

            if node.tokenName == "T_SYMBOL":
                #format.setForeground(Qt.black)
                if node.code in ["self", "super"]:
                    format.setFontWeight(QFont.Bold)
                
            if node.tokenName in ["T_LITERAL", "T_NUMBER"]:
                format.setForeground(Qt.magenta)
                
            if node.tokenName == "T_COMMENT":
                format.setForeground(Qt.gray)
                
            return format
                
            # white, black, red, darkRed, green, darkGreen, blue, darkBlue, cyan, darkCyan, magenta, 
            # darkMagenta, yellow, darkYellow, gray, darkGray, lightGray, transparent, color0, color1
                
        elif node.type == "identifier":
            if node.parent != None and node.parent.type == "raise":
                format = QTextCharFormat()
                format.setForeground(Qt.red)
                return format
                
            if node.parent != None and node.parent.type == "call":
                format = QTextCharFormat()
                format.setForeground(Qt.blue)
                return format
                
            if node.parent != None and node.parent.type in ["function", "class"]:
                format = QTextCharFormat()
                format.setForeground(Qt.darkYellow)
                format.setFontWeight(QFont.Bold)
                return format
                
        return None

    def populateFileContext(self, context):
    
        filePath = context.filePath[len(context.projectFolder)+1:]
        
        # TODO: This assumes that the project-root is also the python-path root
        pythonPath = filePath.replace(".py", "").replace("/", ".")
        
        # print(filePath, pythonPath)
    
        for classNode in context.syntaxTree.find("class"):
            classBlock = classNode.next()
            classDef = ClassDef(
                classNode.find("identifier")[0].code, 
                namespace = pythonPath,
                parents=list(map(lambda a: a.code, classNode.find("tuple-element"))),
                node=classNode,
                block=classBlock
            )
            context.addClass(classDef)
            
            for methodNode in classBlock.find("function"):
                print(methodNode)
                
                methodArguments = []
                for element in methodNode.find("tuple-element"):
                    methodArguments.append(element.code)
                
                MethodDef(
                    classDef,
                    methodNode.children[1].code,
                    methodArguments
                )
                
        for functionNode in context.syntaxTree.find("function"):
            functionArguments = []
            for element in functionNode.find("tuple-element"):
                functionArguments.append(element.code)
            context.addFunction(FunctionDef(
                functionNode.children[1].code,
                functionArguments
            ))
        

File: java/src/main/java/de/addiks/einsicht/languages/PythonLanguage.java
package de.addiks.einsicht.languages;

import de.addiks.einsicht.Hub;
import de.addiks.einsicht.abstract_syntax_tree.ASTNode;
import de.addiks.einsicht.abstract_syntax_tree.patterns.*;
import de.addiks.einsicht.abstract_syntax_tree.selectors.ASTSelector;
import de.addiks.einsicht.semantics.ClassHead;
import de.addiks.einsicht.semantics.FunctionHead;
import de.addiks.einsicht.semantics.Semantic;
import de.addiks.einsicht.abstract_syntax_tree.tokens.Token;
import de.addiks.einsicht.abstract_syntax_tree.tokens.TokenMatcher;
import de.addiks.einsicht.abstract_syntax_tree.tokens.matchers.DirectTokenMatcher;
import de.addiks.einsicht.abstract_syntax_tree.tokens.matchers.KeywordsTokenMatcher;
import de.addiks.einsicht.abstract_syntax_tree.tokens.matchers.LiteralTokenMatcher;
import de.addiks.einsicht.abstract_syntax_tree.tokens.matchers.RegexMatcher;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

public class PythonLanguage extends Language {
    public PythonLanguage(Hub hub) {
        super(hub);
    }

    @Override
    public String name() {
        return "Python";
    }

    @Override
    public List<TokenMatcher> tokenMatchers() {
        return List.of(
                new KeywordsTokenMatcher(List.of(
                        "from", "import", "assert",
                        "def", "pass", "class", "lambda",
                        "if", "elif", "else", "not",
                        "return", "break", "continue",
                        "raise", "except", "try", "finally",
                        "while", "for", "match", "case",
                        "True", "False", "None",
                        "and", "or", "xor", "not",
                        "del", "as", "in", "breakpoint"
                )),
                new RegexMatcher(Pattern.compile("\\n( +)?(?=\\n)"), "T_INDENTATION_SUPERFLOUS"),
                new RegexMatcher(Pattern.compile("\\n( +)?"), "T_INDENTATION"),
                new RegexMatcher(Pattern.compile("\\s+"), "T_WHITESPACE"),
                new DirectTokenMatcher("@", "T_DECORATOR"),
                new RegexMatcher(Pattern.compile("[a-zA-Z_][a-zA-Z0-9_]+"), "T_SYMBOL"),
                new RegexMatcher(Pattern.compile("[0-9_]+(\\.[0-9]*)?"), "T_NUMBER"),
                new RegexMatcher(Pattern.compile("#([^\\n]*)"), "T_COMMENT"),
                new DirectTokenMatcher(List.of(".", ",", "(", ")", "[", "]", "{", "}", ":", "="), "T_SPECIAL_CHAR"),
                new DirectTokenMatcher(List.of("=", "+=", "-=", "*=", "/="), "T_ASSIGN"),
                new DirectTokenMatcher(List.of("+", "-", "*", "/"), "T_OPERATOR"),
                new LiteralTokenMatcher('\'', "T_LITERAL"),
                new LiteralTokenMatcher('"', "T_LITERAL")
        );
    }

    @Override
    public boolean isNodeRelevantForGrammar(ASTNode node) {
        if (node instanceof Token token) {
            return !List.of("T_WHITESPACE", "T_COMMENT", "T_INDENTATION").contains(token.tokenName());
        }
        return true;
    }

    @Override
    public List<NodePattern> grammar() {
        LateDefinedASTPattern expression = new LateDefinedASTPattern("expression");

        NodeSequence identifier = new NodeSequence("identifier", List.of(
                new TokenNodePattern("T_SYMBOL"),
                new RepeatingNode("identifier-path", new NodeSequence("identifier-element", List.of(
                        new TokenNodePattern("."),
                        new TokenNodePattern("T_SYMBOL")
                )), true)
        ));

        NodeSequence importFrom = new NodeSequence("import-from", List.of(
                new TokenNodePattern("T_FROM"),
                new RepeatingNode("import-from-parent", new TokenNodePattern("."), true),
                identifier
        ));

        NodeSequence importStatement = new NodeSequence("import", List.of(
                new OptionalNode(importFrom),
                new TokenNodePattern("T_IMPORT"),
                identifier
        ));

        NodeSequence tuple = new NodeSequence("tuple", List.of(
                new TokenNodePattern("("),
                new RepeatingNode("tuple-content", new NodeSequence("tuple-element", List.of(
                        expression,
                        new OptionalNode(new TokenNodePattern(","))
                )), true),
                new TokenNodePattern(")")
        ));

        NodeSequence list = new NodeSequence("list", List.of(
                new TokenNodePattern("["),
                new RepeatingNode("list-content", new NodeSequence("list-element", List.of(
                        expression,
                        new OptionalNode(new TokenNodePattern(","))
                )), true)
        ));

        NodeSequence call = new NodeSequence("call", List.of(
                identifier,
                tuple
        ));

        NodeSequence raiseDef = new NodeSequence("raise", List.of(
                new TokenNodePattern("T_RAISE"),
                identifier,
                tuple
        ));

        NodeSequence decorator = new NodeSequence("decorator", List.of(
                new TokenNodePattern("@"),
                identifier,
                new OptionalNode(tuple)
        ));

        NodeSequence operation = new NodeSequence("operation", List.of(
                expression,
                new TokenNodePattern("T_OPERATOR"),
                expression
        ));

        NodeSequence assignment = new NodeSequence("assignment", List.of(
                expression,
                new TokenNodePattern("="),
                expression
        ));

        NodeSequence classDefinition = new NodeSequence("class", List.of(
                new TokenNodePattern("T_CLASS"),
                identifier,
                tuple,
                new TokenNodePattern(":")
        ));

        NodeSequence functionDefinition = new NodeSequence("function", List.of(
                new TokenNodePattern("T_DEF"),
                identifier,
                tuple,
                new TokenNodePattern(":")
        ));

        NodeBranch bool = new NodeBranch("boolean", List.of(
                new TokenNodePattern("T_TRUE"),
                new TokenNodePattern("T_FALSE")
        ));

        NodeBranch expressionElement = new NodeBranch("expression-element", List.of(
                identifier,
                call,
                bool,
                list,
                tuple,
                new TokenNodePattern("T_LITERAL"),
                new TokenNodePattern("T_NUMBER")
                // operation # TODO: Infinite loop? expr => operation => expression => ... HOW?!
        ));

        NodeBranch expressionSuffix = new NodeBranch("expression-suffix", List.of(list));

        expression.definePattern(new NodeSequence("expression", List.of(
                expressionElement,
                new RepeatingNode("expression-suffixes", expressionSuffix, true)
        )));

        return List.of(
                importStatement,
                importFrom,
                identifier,
                tuple,
                list,
                bool,
                classDefinition,
                functionDefinition,
                raiseDef,
                call,
                decorator,
                operation,
                assignment,
                expression,
                expressionElement
        );
    }

    @Override
    public Map<String, Semantic.Factory> semantics() {
        Map<String, Semantic.Factory> semantics = new HashMap<>();

        semantics.put("class", new ClassHead.Factory(
                ASTSelector.of("class/identifier")
        ));

        semantics.put("function", new FunctionHead.Factory(
                ASTSelector.of("function/identifier")
        ));

        return semantics;
    }

    @Override
    public Object stylesheet() {
        return null;
    }
}
